### ISSUE TITLE
Should we set default margins on 'p' tag in Basis?

### ISSUE THREAD

**Issue by stpaultim** [backdrop/backdrop-issues#4166]
**Description of the bug**
In some ways this might be a really small fringe issue, or it might be something I deal with quite a bit. 

In further exploring this issue (https://github.com/backdrop/backdrop-issues/issues/4165), I realized that if you create an inline label on a body field (an unusual use case) in Basis, the alignment is pretty weird. 

![Your_first_post____Test](https://user-images.githubusercontent.com/3144571/67643523-788a3200-f8e6-11e9-8630-daa9d369ac39.png)

This is NOT a problem in Bartik. Turns out that Bartik applies the following CSS to any paragraph tag:

```
p {
    margin: 0 0 1.2em;
}
```

Basis does not apply any properties to the p tag, leaving it to the ```user agent stylesheet``` (browser defaults). In all the browsers I tested, the user agent stylesheet puts a top and bottom margin on paragraphs. 

This means that anytime you put a paragraph next to another element in two columns, they will not align nicely on top, because the paragraph has this top margin. As I think about it, I think I battle this quite a bit, when I try to do things like put a body field block in one column and an image block in a neighboring column. 

![Your_first_post____Test](https://user-images.githubusercontent.com/3144571/67643779-154dcf00-f8e9-11e9-8a58-02502ca1d411.png)

As themers, I expect that most of us are used to making an adjustment for this. But, I would expect that this should work better out of the box for users using Basis. 

I'm going to recommend that we specify that there be no top padding on paragraph tags. 

This is also an issue with h1, h2, etc tags. Both Bartik and Basis specifically do give these items a margin on top. Let's save that for another issue. 

I wonder what others think about this? 

- - - - - - -

PR by @stpaultim = https://github.com/backdrop/backdrop/pull/2954

**Comment by stpaultim** [comment]
I went ahead and created a PR that adds margins to ```p``` tag in Basis. 

I am not sure if this is the right solution, but thought it could not hurt to put this PR on the table for feedback and testing. 

If there is some risk that this might break existing sites, I would consider that it might be time to create a new theme in core that fixes these kinds of issues.

**Comment by olafgrabienski** [comment]
> I wonder what others think about this?

Interesting: When I read this issue I was quite sure that I've changed the CSS for `<p>` tags on most of 'my' *Basis* sites. But surprisingly I changed it only on one of them.

Looks like I'm happy without the explicit margin by the theme. However, usually I make a sub-theme of *Basis*, so I know I can override the browser defaults anytime.

**Comment by Graham-72** [comment]
I have a number of sites that use a custom sub-theme of Basis and I am worried by the prospect of introducing new style settings in core. I think we would need to have some way of notifying us of such a change in advance of its release so that the sub-theme style-sheet can be modified if necessary.
However, I have checked the impact of this proposed PR on one of my sites and it makes very little difference and would make even less with a margin of 1.1 rather than 1.2 em.

**Comment by olafgrabienski** [comment]
> I think we would need to have some way of notifying us of such a change in advance of its release so that the sub-theme style-sheet can be modified if necessary.

Do you think it's enough to mention such modifications in the release notes?

> I have checked the impact of this proposed PR on one of my sites and it makes very little difference and would make even less with a margin of 1.1 rather than 1.2 em.

User agents declare `margin-block-start` and `margin-block-end` of `1em`, so I'd suggest `margin: 0 0 1em;`. That said, I'm not sure if we should use `rem`, that's the unit which *Basis*  uses.

**Comment by stpaultim** [comment]
I share the concern about effecting other sites and I think that there are other issues like this that need addressing, which is why I also raised the topic of a new theme for Backdrop CMS in which we try to address these kinds of issues (just throwing out ideas for now, I am not convinced it is a good/necessary idea):

https://github.com/backdrop/backdrop-issues/issues/4167

@olafgrabienski - Disregarding my specific proposal, what do you think of the original issue/screenshot of how a field with a <p> tag in it does not align well (inline) with it's label. Does this need a fix (I think so)? And are there other solutions?

**Comment by olafgrabienski** [comment]
> ... in fact it appears that Basis seems to use both REM and EM and I see that most all padding and margins are using EM.

You're right, both REM and EM are used, and EM seems to be more frequent for margins and paddings; I should have checked that before.

**Comment by olafgrabienski** [comment]
> Disregarding my specific proposal, what do you think of the original issue/screenshot of how a field with a `<p>` tag in it does not align well (inline) with it's label. Does this need a fix (I think so)? And are there other solutions?

In my opinion, an inline label for a body field or another long text field is an edge case which doesn't need to be fixed *urgently*. However, there will be real use cases, e.g. a data-rich content type with many fields which lists every field with a label to give a good overview.

So after all a fix would be welcome but it may be hard to address the issue without affecting other sites / use cases. I have two (general) ideas for this issue and comparable ones in mind:

- Provide an option to add custom CSS in Backdrop core.
- Make it super easy to make a sub-theme.

(There are probably already respective feature requests somewhere.)

**Comment by stpaultim** [comment]
I agree that this is an edge case and not urgent, but it's interesting to me because it addresses this larger issue of how to fix Basis without breaking existing sites.

Issue to add sub-theme to basis: https://github.com/backdrop/backdrop-issues/issues/3914

**Comment by olafgrabienski** [comment]
>> - Provide an option to add custom CSS in Backdrop core.
>> - Make it super easy to make a sub-theme.

> Issue to add sub-theme to basis: #3914 

Thanks @stpaultim! I've also searched for a "Custom CSS" feature request but it doesn't seem to exist so far. What I've found: a meta issue ([Make subtheming easier](https://github.com/backdrop/backdrop-issues/issues/1829)) where, amongst other things, we talked about [making theme CSS overrides easier](https://github.com/backdrop/backdrop-issues/issues/1829#issuecomment-314055011).

A while after that discussion, the *CSS Injector* module was ported to Backdrop, which is great, but I'd still like to have a nicely integrated *Custom CSS* option in core (WordPress does, btw). @stpaultim Would the ability to provide custom CSS be a good approach for your question ("how to fix Basis without breaking existing sites")?

**Comment by stpaultim** [comment]
@olafgrabienski - I don't think that the ability for a user to add CSS through the UI is a sufficient fix for problems like this one. I believe that we should find a way in core to fix this issue that does not require a user to inject css solutions on their own. 

But, I do support the idea of providing CSS injector features in core as an option.

**Comment by jenlampton** [comment]
I was hoping we could talk about this in the meeting today, but forgot :( 

Just brainstorming here... What if we reserved all CSS changes like this for minor versions only, and with each minor version added an additional stylesheet to Basis. Something like `/core/themes/basis/css/updates/14.css`. When you install Backdrop, we save that minor version somewhere secretly, and then your site would get only the stylesheets that were in core when you first installed.

How would we make this work...  Well, we could add the CSS in `preprocess_page()`...  if we recorded the minor version of Backdrop that were first installed somewhere, we could pull that out here to test. Maybe something like...

```php
function basis_preprocess_page(&$variables) {
  $path = backdrop_get_path('theme', 'basis');
  $install_version = config_get('system.core', 'installed_minor_version');
  for ($i = 1; $i <= $install_version; $i++) {
    $filename = $path .'/css/updates/' . $i . '.css';
    if (file_exists($filename)) {
      backdrop_add_css($filename);
    }
  }
  ...
```

**Comment by stpaultim** [comment]
I didn't bring it up during meeting, because it was already running late and it's not urgent. Your solution makes rough sense to me and addresses the larger issue here: https://github.com/backdrop/backdrop-issues/issues/4167
